### ISSUE TITLE
DPO et RLHF : aligner un modèle sans douleur

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#491]
---
lede: Aligner un modèle sur les préférences humaines opposait RLHF, puissant mais lourd, à DPO, simple mais fragile — et 2026 tranche moins par la théorie que par le coût opérationnel.
lede_en: Aligning a model on human preferences pitted RLHF, powerful but heavy, against DPO, simple but fragile — and 2026 decides less on theory than on operational cost.
title_en: DPO and RLHF: aligning a model painlessly
---

## Pourquoi aligner, et sur quoi

Un modèle pré-entraîné prédit le token suivant ; il ne sait pas être *utile*, *honnête* ou *inoffensif*. Le fine-tuning supervisé (SFT) lui apprend le format des réponses, mais pas les préférences fines : quelle réponse est *meilleure* entre deux qui sont toutes deux correctes ? C'est le rôle de l'alignement par préférences.

Le principe commun à RLHF et DPO : partir de **paires de préférences**. On montre au modèle deux réponses à un même prompt, et un humain (ou un modèle) indique laquelle est préférée. Le modèle apprend à produire davantage la réponse « gagnante » et moins la « perdante ».

- On ne dit pas au modèle *la bonne réponse*, mais *quelle réponse est préférée à quelle autre*.
- La donnée est un tripletet `(prompt, choisie, rejetée)`.
- La qualité de l'alignement dépend d'abord de la qualité de ces préférences.

## RLHF : puissant mais une usine à gaz

Le RLHF (Reinforcement Learning from Human Feedback) fut la méthode fondatrice. Elle se déroule en trois temps :

1. **SFT** — fine-tuning supervisé de base.
2. **Reward model** — on entraîne un second modèle à prédire la préférence humaine (donner un score à une réponse).
3. **RL (PPO)** — on optimise le modèle par renforcement pour maximiser ce score, avec une pénalité KL qui l'empêche de trop s'éloigner du modèle de départ.

Le résultat est excellent mais le pipeline est **lourd et instable** :

- Il faut **deux modèles en mémoire** (policy + reward), plus une copie de référence.
- **PPO est capricieux** — sensible aux hyperparamètres, sujet au *reward hacking* (le modèle exploite les failles du reward model au lieu de vraiment s'améliorer).
- L'infrastructure RL est complexe à opérer et à déboguer.

## DPO : la même cible, sans le RL

Le DPO (Direct Preference Optimization) est l'insight qui a simplifié le domaine : on peut atteindre le même objectif **sans reward model ni RL**, par une simple perte de classification sur les paires de préférences. Mathématiquement, DPO montre que la politique optimale du RLHF a une forme fermée, qu'on optimise directement.

```python
# Perte DPO, schématique
import torch.nn.functional as F

def dpo_loss(logp_chosen, logp_rejected,
             ref_logp_chosen, ref_logp_rejected, beta=0.1):
    # écart de log-prob entre modèle courant et référence
    pi   = logp_chosen - logp_rejected
    ref  = ref_logp_chosen - ref_logp_rejected
    return -F.logsigmoid(beta * (pi - ref)).mean()
```

Les avantages sont concrets :

- **Un seul entraînement**, pas de reward model séparé, pas de boucle RL.
- **Stable et reproductible** — c'est une descente de gradient classique.
- **Accessible** — tient sur des budgets bien plus modestes.

Le prix : DPO est **plus sensible à la qualité des paires** et peut sur-optimiser (dégrader la diversité, devenir trop prudent) si `beta` et les données sont mal calibrés.

## Le paysage de 2026

La réalité pratique aujourd'hui :

- **DPO est le défaut** pour la plupart des équipes — le rapport résultat/effort est imbattable, et des variantes (IPO, KTO, ORPO) corrigent ses fragilités.
- **RLHF garde l'avantage** sur les gros laboratoires visant le sommet de l'alignement, où le reward model apporte une supervision plus riche et l'exploration RL découvre des comportements que les paires figées ne montrent pas.
- **Le feedback IA** (RLAIF) remplace de plus en plus l'annotation humaine coûteuse : un modèle juge génère les préférences, ce qui déplace le problème vers la fiabilité du juge.
- La donnée de préférence reste **le vrai goulot** — bien plus que le choix d'algorithme.

## Verdict

Pour 90 % des cas en 2026, **commencez par DPO** : simple, stable, abordable, et suffisant pour aligner un modèle sur un ton, une politique de refus, un style. Ne montez à RLHF que si vous poursuivez le dernier pourcent d'alignement avec les moyens d'un labo — et acceptez son coût opérationnel et son risque de reward hacking. Mais retenez surtout que le choix d'algorithme est secondaire : la qualité, la cohérence et la couverture de vos **paires de préférences** déterminent le résultat bien plus que DPO contre RLHF. Investissez dans les données, pas dans la hype de la méthode.

## Why align, and on what

A pre-trained model predicts the next token; it doesn't know how to be *helpful*, *honest* or *harmless*. Supervised fine-tuning (SFT) teaches it the format of answers, but not fine preferences: which answer is *better* between two that are both correct? That's the role of preference alignment.

The principle common to RLHF and DPO: start from **preference pairs**. You show the model two answers to the same prompt, and a human (or a model) indicates which is preferred. The model learns to produce more of the "winning" answer and less of the "losing" one.

- You don't tell the model *the right answer*, but *which answer is preferred over which*.
- The data is a triplet `(prompt, chosen, rejected)`.
- Alignment quality depends first on the quality of those preferences.

## RLHF: powerful but a contraption

RLHF (Reinforcement Learning from Human Feedback) was the founding method. It unfolds in three stages:

1. **SFT** — base supervised fine-tuning.
2. **Reward model** — you train a second model to predict human preference (score an answer).
3. **RL (PPO)** — you optimize the model by reinforcement to maximize that score, with a KL penalty keeping it from drifting too far from the starting model.

The result is excellent but the pipeline is **heavy and unstable**:

- You need **two models in memory** (policy + reward), plus a reference copy.
- **PPO is finicky** — sensitive to hyperparameters, prone to *reward hacking* (the model exploits the reward model's flaws instead of genuinely improving).
- RL infrastructure is complex to operate and debug.

## DPO: the same target, without the RL

DPO (Direct Preference Optimization) is the insight that simplified the field: you can reach the same objective **without a reward model or RL**, through a simple classification loss over preference pairs. Mathematically, DPO shows that RLHF's optimal policy has a closed form, which you optimize directly.

```python
# DPO loss, schematic
import torch.nn.functional as F

def dpo_loss(logp_chosen, logp_rejected,
             ref_logp_chosen, ref_logp_rejected, beta=0.1):
    # log-prob gap between current model and reference
    pi   = logp_chosen - logp_rejected
    ref  = ref_logp_chosen - ref_logp_rejected
    return -F.logsigmoid(beta * (pi - ref)).mean()
```

The advantages are concrete:

- **A single training run**, no separate reward model, no RL loop.
- **Stable and reproducible** — it's classic gradient descent.
- **Accessible** — fits on far more modest budgets.

The price: DPO is **more sensitive to pair quality** and can over-optimize (degrade diversity, become too cautious) if `beta` and the data are poorly calibrated.

## The 2026 landscape

The practical reality today:

- **DPO is the default** for most teams — the result/effort ratio is unbeatable, and variants (IPO, KTO, ORPO) fix its fragilities.
- **RLHF keeps the edge** at large labs chasing the peak of alignment, where the reward model provides richer supervision and RL exploration discovers behaviors frozen pairs don't show.
- **AI feedback** (RLAIF) increasingly replaces costly human annotation: a judge model generates preferences, which shifts the problem to judge reliability.
- Preference data remains **the real bottleneck** — far more than the algorithm choice.

## Verdict

For 90% of cases in 2026, **start with DPO**: simple, stable, affordable, and enough to align a model on a tone, a refusal policy, a style. Move up to RLHF only if you're chasing the last percent of alignment with a lab's resources — and accept its operational cost and reward-hacking risk. But above all, remember the algorithm choice is secondary: the quality, consistency and coverage of your **preference pairs** determine the outcome far more than DPO versus RLHF. Invest in the data, not in the method's hype.
